Chap_API_Sharing_Basics.tex

```diff
@@ -0,0 +1,630 @@
+which includes the type of the value's data.  
+Each key is unique within the context or domain in which it exists.
+The context or domain of a key in PMIx is called a data realm.  
+%% The pmix_info_t type (See \ref{chap:introduction:portability:attributes}), 
+%% is also built on key-values pairs.  
+%% However, pmix_info_t is used for influencing API calls and is not used for sharing 
+%% information between processes.  
+This chapter describes how key-value pairs are made available and accessed by clients.
+
+A key-value pair has a scope defined by its data realm.  
+PMIx defines data realms for sessions, jobs, applications, nodes, processes and a special 
+realm for publish/lookup information.   
+These are explained more fully in chapters \ref{chap:api_rsvd_keys} and \ref{chap:pub}.
+
+The source or provider of key-value pairs can be the host environment or a \ac{PMIx} client.  
+The host environment can provide key-values pairs in any realm except for the publish/lookup realm.
+%%% Can we have a separate effort to change this restriction on publish/lookup for consistency?
+%%% Would this get into too many other issues?
+PMIx clients can only provide key-values pairs within their own process realm and the publish/lookup realm.
+The value of a key is not necessarily static and can be modified by the original provider of the key.
+
+\ac{PMIx} classifies keys as either reserved or non-reserved.
+
+Only keys provided by a host environment or \ac{PMIx} implementation can begin with \code{"pmix"} and are called reserved keys.  
+Host environments and \ac{PMIx} implementations must only provide keys that begin with \code{"pmix"} to avoid conflicts with keys that are provided by clients.
+Reserved keys are documented in chapter \ref{chap:api_rsvd_keys}.
+Each standardized reserved key has an associated macro that is documented along with its description.
+Likewise clients can only provide keys that do not begin with \code{"pmix"} to avoid conflicting with keys provided by the host environment or \ac{PMIx} implementation.   
+The string representation and the macro are equivalent methods to refer to a key.
+Throughout this document, the macro is used rather than its string value to refer to reserved keys.
+Clients cannot create reserved keys, nor can they modify the value of reserved keys.
+The way reserved keys are accessed by clients 
+is different than for non-reserved keys as explained in \ref{chap:api_rsvd_keys}.
+
+Host environments are not required to provide all keys that are documented in the standard.  
+However, when providing keys that are documented, they must adhere to the behavior defined by the standard for reserved keys.  
+An implementation cannot provide a reserved key that behaves differently than what is in the standard.  
+Implementations are encouraged to standardize any non-standardized, reserved keys that they provide to avoid conflicting with other implementations or efforts to standardize the same key.
```

Comment:
It is implied here, but you might want to explicitly state that an implementation can define reserved keys outside of the standard. However, host environments are not permitted to do so as they can conflict with implementations.

Reply:
You said earlier that you thought it was a bad idea that we allow implementations to define reserved keys.  Do you really mean reserved keys here or do you mean keys that begin with something other than "pmix_", but are made available at the start of client execution?   Right now (both in current standard and this proposed change), reserved key means any key that starts with "pmix_".

Reply:
I'm not wild about implementations doing it. However, the sentence implies that it is okay - we just "encourage" them to file an RFC about them. Obviously, we can't stop anyone from defining such a key - heck, I do it in OpenPMIx just to move things along. So maybe this is the best we can do? My main concern really was the line that followed and to ensure that hosts do not define reserved keys.